### ISSUE TITLE
Issue with projection and geopandas interface

### ISSUE THREAD

**Issue by adelarue** [gboeing/osmnx#677]
I am building a network using `graph_from_gdfs`, after creating `nodes` and `edges` geodataframes myself. The nodes and edges are in a projected coordinate system (i.e. not raw latitude-longitude). However, when I create the new graph, then extract the nodes and edges using `graph_to_gdfs`, it looks like the nodes have been "un-projected" (back to latitude-longitude), but the edges have not - they remain in the same coordinate system. Furthermore, the created geodataframe is inconsistent - the coordinates are clearly latitudes and longitudes, but the associated projection system indicates otherwise, e.g.

<img width="602" alt="Screen Shot 2021-03-14 at 3 19 47 PM" src="https://user-images.githubusercontent.com/7780501/111081327-2943a200-84d9-11eb-8afc-c636a308903a.png">

_but_

<img width="488" alt="Screen Shot 2021-03-14 at 3 20 04 PM" src="https://user-images.githubusercontent.com/7780501/111081348-3d879f00-84d9-11eb-82ea-b076de99e5e3.png">

Minimum working example:

```python
import osmnx as ox
import geopandas as gp
# get an example network
network = ox.graph_from_place("Sausalito, CA, USA", network_type='drive', simplify=False)
# extract corresponding geodataframes
nodes, edges = ox.graph_to_gdfs(network)
# project geodataframes to "flat" local coordinates
edges = edges.to_crs("epsg:7131")
nodes = nodes.to_crs("epsg:7131")
# recreate network from projected data
network = ox.graph_from_gdfs(nodes, edges)
# extract geodataframes again. The nodes have been unprojected, but the edges have not
# additionally, the nodes geodataframe is now invalid (crs does not match coordinates)
nodes, edges = ox.graph_to_gdfs(network)
```

**Comment by gboeing** [comment]
Thanks @adelarue. The issue is that OSMnx uses the nodes' `x` and `y` attributes to construct their geometry when you use `graph_to_gdfs`.  The easiest fix would be to project the graph directly with OSMnx:

```python
G1 = ox.graph_from_place("Sausalito, CA, USA", network_type='drive', simplify=False)
G1 = ox.project_graph(G1, "epsg:7131")
nodes1, edges1 = ox.graph_to_gdfs(G1)
G2 = ox.graph_from_gdfs(nodes1, edges1)
nodes2, edges2 = ox.graph_to_gdfs(G2)
nodes2
```

This should give you the result you're looking for. It works because OSMnx keeps the graph's nodes' `x` and `y` attributes in sync with their `geometry` when you project this way, and preserves the original `x` and `y` coordinates as new `lat` and `lon` attributes. This is the method I'd recommend.

However, the use you documented results in surprising behavior which should probably be fixed. One fix might be changing `graph_to_gdfs` so that if the graph nodes have a `geometry` attribute, then use that as the new GeoDataFrame's geometry column; otherwise, convert the nodes' `x` and `y` attributes to Points for the geometry column (as currently happens).

**Comment by adelarue** [comment]
Makes sense - I didn't realize the projection relied on `x` and `y` over `geometry`. I guess it is also bad practice on my part to change the geometry but leave `x` and `y` floating around unchanged.

I wonder if the fix should actually happen earlier in the process? When I inspect the `network` object at the end of the MWE, it looks like the geometry column is in projected coordinates and the x and y columns are in latitude-longitude. I don't think an OSMnx graph should ever be in such a state (for example, it also breaks plotting functions, because the edges are projected correctly, while the nodes are in a "half-projected" state).

Would it be possible to impose a match between `geometry` and `x` and `y` for the nodes as an invariant, and throw an error when `graph_from_gdfs` receives a `nodes` data frame that does not comply?

**Comment by gboeing** [comment]
@adelarue good point. Yes possibly. It's worth considering how the [projection](https://github.com/gboeing/osmnx/blob/6fce4c8eaae69ccbd841f217544011d367e9847b/osmnx/projection.py#L136-L138) module keeps `x` and `y` in sync with the geometry, and drops the therefore superfluous `geometry` attribute itself before rebuilding the graph.

Perhaps, along the lines of your suggestion, the `graph_from_gdfs` function should:

  1. check if the nodes have a `geometry` column
  2. if so, verify that its values match those of the `x` and `y` columns
  3. if they match, drop the `geometry` column so it doesn't end up in the resulting graph as a superfluous attribute
  4. if they don't match, throw an error.

Thoughts?

**Comment by adelarue** [comment]
I like this solution! The benefit is that it prevents "corrupt" data from ever entering the OSMnx graph.

In (1) I assume you mean to check if the `nodes` GeoDataFrame has a `geometry` attribute, since I think the column itself could be called something else.

Checking for a match between (`x`, `y`) and the `geometry` column might have some weird edge cases (what if someone represents a node as a really small line segment or polygon?) Throwing an error still makes sense because these unorthodox geometries would be at risk of being deleted under the hood by calling `graph_to_gdfs`.

Thanks for the thoughtful and quick responses!

**Comment by gboeing** [comment]
> I assume you mean to check if the nodes GeoDataFrame has a geometry attribute

Right, yes, an *attribute* rather.

Would you mind taking a look at the proposed fix in #678? I decided to go with a user warning + better documentation to hopefully convey to the user how to use it and if there's an impending problem.
